x-pack/plugins/lens/public/datatable_visualization/visualization.tsx

```diff
@@ -22,6 +22,10 @@ export interface LayerState {
 
 export interface DatatableVisualizationState {
   layers: LayerState[];
+  sorting?: {
+    columnId: string | undefined;
+    direction: 'asc' | 'desc';
+  };
```

Comment:
Why is the sorting attached to the top level of the state, instead of attached to a specific layer?

Also, what happens in these two cases?

1. A sorted column is deleted- is the state correctly updated?
2. The columnId for a sorted column is reused but the function has changed. It's still sorted, but should it be?

Reply:
Yes, I've thought about that and discussed a bit with @flash1293 about.
We agreed to have sorting on a visualization level for now, as tables can only have one layer.
When we'll extend tables support for multiple layer this feature will be revisited, as probably other changes are planned to happen to this code meanwhile.

Reply:
No strong feeling about the sorting state - @dej611 and me discussed this but I think it's fine in both places. I wasn't aware we are supporting multiple layers in the first place, what's the reasoning for this? (it seems like it's not possible for the user to actually configure this and even if they did, it would just show the first table)

> The columnId for a sorted column is reused but the function has changed. It's still sorted, but should it be?

IMHO it should still be sorted as long as it's the same dimension/column. If the dimension gets deleted completely, the sorting state should be deleted along with it.

Reply:
> * A sorted column is deleted- is the state correctly updated?
Probably the state in this case will remain "dirty" but the final result should be correct anyway as it relies on the `columnId`. I'll try to address this scenario.

> * The columnId for a sorted column is reused but the function has changed. It's still sorted, but should it be?

I wasn't aware of recycled `ids`. I'll take care for that too.

Reply:
Had a first look at this implementation: the formatters are available only at render time as dependency, not at expression time, going in conflict on the CSV sorted downloadable content.

I'll investigate on possible solutions. In case I'd rather prioritise this over the CSV sorted downloadable content.

Reply:
Ah, good point. Agreed, we can open a separate issue for CSV sorted downloadable content. It's fixable by passing down the reference to the format factory from data plugin from the `setup` method of the `DatatableVisualization` service class in `x-pack/plugins/lens/public/datatable_visualization/index.ts` and wrapping the datatable visualization definition into a getter (`getDatatableVisualizationDefinition(data: DataPublicPluginStart)`)